dlib/global_optimization/global_function_search.h

```diff
@@ -13,6 +13,19 @@
 
 namespace dlib
 {
+    class stop_condition {
+      public:
+      template <class Function>
+      explicit stop_condition(Function&& f): check(std::forward<Function>(f)) {}
+      // returns true if the search should terminate early
+      bool operator()(double v) {
+        return check(v);
+      }
+      std::function<bool(double)> check;
+    };
+
+    // a default stop condition that never terminates early
+    extern stop_condition default_stop_condition;
```

Comment:
This shouldn't be needed.  You can make the functions just take a generic lambda.  See my other comment for more details.

Reply:
I thought about a generic lambda, but I favored a separate class for a C++11 compatible way to enforce the type signature on the function-object and because it makes the call site more self documenting.  I also wasn't sure how well function arguments of template type worked with defaults in older compilers that dlib supports.

I can change this if needed.

Reply:
A generic lambda is the idiomatic way to do this in c++11 or later.  And using std::function when extra type emphasis is called for, which I think is the case here.